dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1094 @@
+the HWC x RS matrix, consecutive rows map to consecutive memory addresses (with the exception of
+positions when we cross H, W extents and border remapping takes place). This implementation melds
+the W and C extents to utilize this property. Additionally, the implementation that uses shared
+memory loads the inputs in blocks with extents corresponding to (D, )H, W * C extents to account
+for the fact that spatailly close products will reuse some of the inputs.
+*/
+template <typename SampleDescT, typename BlockSetupFactory, typename InLoaderFactory,
+          typename GridSetupT>
+__global__ void filter(const SampleDescT* __restrict__ descs, BlockSetupFactory block_setup_factory,
+                       InLoaderFactory in_loader_factory, GridSetupT grid_setup) {
+  extern __shared__ char shm[];
+  int sample_idx = grid_setup.sample_idx();
+  for (int sample_idx = grid_setup.sample_idx(); sample_idx < grid_setup.num_samples();
+       sample_idx += grid_setup.sample_dim()) {
+    auto sample_desc = descs[sample_idx];
+    const auto& block_setup = block_setup_factory(sample_idx);
+    const auto& log_block_extents = sample_desc.shape.log_block_extents;
+    auto block_start = grid_setup.block_idx() * log_block_extents;
+    auto process_sample = [&](const auto& out_extents, const auto& out_strides,
+                              const auto& out_frame_stride) {
+      if (any_coord(block_start >= out_extents)) {
+        return;  // early exit to avoid all the setup only to do nothing in the stride loop
+      }
+      auto grid_size = grid_setup.grid_dim() * log_block_extents;
+      const auto& in_loader = in_loader_factory(sample_idx);
+      if (sample_desc.shape.in_workspace_extents.x) {
+        using In = typename SampleDescT::In;
+        int* idx_workspace = reinterpret_cast<int*>(shm);
+        In* in_workspace = reinterpret_cast<In*>(shm + sample_desc.shape.in_workspace_offset);
+        with_shm_conv(
+            sample_desc, in_loader, block_setup, idx_workspace, in_workspace, [&](auto&& conv) {
+              stride_grid(block_start, grid_size, out_extents, out_strides, out_frame_stride, conv);
+            });
+
+      } else {
+        with_direct_conv(sample_desc, in_loader, block_setup, [&](auto&& conv) {
+          stride_grid(block_start, grid_size, out_extents, out_strides, out_frame_stride, conv);
+        });
+      }
+    };
+    // If extents are equal then strides are too, make it explicit as it faster and the usual case
+    // (roi is, for now, only used for ``valid`` mode)
+    if (sample_desc.shape.out_extents == sample_desc.shape.in_extents) {
+      process_sample(sample_desc.shape.in_extents, sample_desc.shape.in_strides,
+                     sample_desc.shape.frame_stride);
+    } else {
+      process_sample(sample_desc.shape.out_extents, sample_desc.shape.out_strides,
+                     sample_desc.shape.out_frame_stride);
+    }
```

Comment:
Is this readibilty only thing?

Reply:
This part is done with the OutputShapeProvider statically.